gadopt/gplates.py

```diff
@@ -0,0 +1,450 @@
+            [pygplates.PointOnSphere(x=coords[i, 0], y=coords[i, 1], z=coords[i, 2], normalise=True)
+             for i in range(np.shape(coords)[0])]
+        )
+
+        # Create a feature containing the multipoint feature, and defined as MeshNode type
+        meshnode_feature = pygplates.Feature(pygplates.FeatureType.create_from_qualified_string('gpml:MeshNode'))
+        meshnode_feature.set_geometry(multi_point)
+        meshnode_feature.set_name('Velocity Mesh Nodes from pygplates')
+
+        output_feature_collection = pygplates.FeatureCollection(meshnode_feature)
+
+        return output_feature_collection
+
+    def _interpolate_seeds_u(self, target_coords):
+        """Interpolates seed velocities onto target coordinates.
+
+        This method generates a KD-tree of seed points with numerical values, then uses
+        this tree to find the nearest neighbors of the target coordinates. It interpolates
+        the velocities of these neighbors to the target points, applying a weighted average
+        based on distance. If any target point is exceptionally close to a seed point,
+        it directly assigns the seed's velocity to avoid division by zero errors.
+
+        Args:
+            target_coords (numpy.ndarray): Array of target coordinates for velocity interpolation.
+
+        Returns:
+            numpy.ndarray: Interpolated velocities at the target coordinates.
+        """
+        # calculate velocities here
+        seeds_u = self._calc_velocities(
+            velocity_domain_features=self.velocity_domain_features,
+            topology_features=self.topology_features,
+            rotation_model=self.rotation_model,
+            age=self.reconstruction_age,
+            delta_t=self.delta_t)
+
+        seeds_u = np.array([i.to_xyz() for i in seeds_u]) *\
+            ((pyGplatesConnector.velocity_dimDcmyr) / self.scaling_factor)
+
+        # if pyGplates does not find a plate id for a point, assings NaN to the velocity
+        # here we make sure we only use velocities that are not NaN.
+        non_nan_values = ~np.isnan(seeds_u[:, 0])
+
+        # generate a KD-tree of the seeds points that have a numerical value
+        tree = cKDTree(data=self.seeds[non_nan_values, :], leafsize=16)
+
+        # find the neighboring points
+        dists, idx = tree.query(x=target_coords, k=self.nneighbours)
+
+        # Use np.where to avoid division by very small values
+        # Replace tiny distances with pyGplatesConnector.epsilon_distance to avoid division
+        # by tiny values while keeping original distances intact for later use
+        safe_dists = np.where(dists < pyGplatesConnector.epsilon_distance, pyGplatesConnector.epsilon_distance, dists)
+
+        # Then, calculate the weighted average using safe_dists
+        res_u = np.einsum(
+            'i, ij->ij',
+            1/np.sum(1/safe_dists, axis=1),
+            np.einsum('ij, ijk ->ik', 1/safe_dists, seeds_u[non_nan_values][idx])
```

Comment:
Just out of curiosity, what is the mathematical operation represented here in the inner call to `einsum`?

Reply:
The inner einsum is multiplying the weighting associated with each point with all three velocity vectors.

Reply:
"i" is the index for number of points, "j" is the index for n number of nearest points, and "k" is the number of coordinates which is 3. This multiplies all three components with the same weight and then sums all the weighted values togeth to give "ik"

Reply:
That is interesting. I gave it a go and indeed:
```python
>>> a = np.ones((2, 3))
>>> b = 2 * np.ones((2, 3, 4))
>>> np.einsum("ij,ijk->ik", a, b)
array([[6., 6., 6., 6.],
       [6., 6., 6., 6.]])
>>> np.sum(np.atleast_3d(a) * b, axis=1)
array([[6., 6., 6., 6.],
       [6., 6., 6., 6.]])
```
I may start using `einsum` more often!